File: Minimum Number_of_Days_to_Disconnect_Island.cpp
class Solution {
    
public:
    int disc[10001], low[10001], parent[10001];
    int flag=2, t=0;
    int apdfs(vector<vector<int>> &adj, vector<int> &vis, int src){
        t++;
        int node=0;
        disc[src]=low[src]=t;
        for(int i:adj[src]){
            if(vis[i]==0){
                vis[i]=1;
                parent[i]=src;
                node++;
                apdfs(adj, vis, i);
                low[src]=min(low[src], low[i]);
                if(node>1 && parent[src]==0)flag= 1;
                if(parent[src]!=0){
                    if(disc[src]<=low[i]){flag= 1;}
                }
            }
            else if(parent[src]!=i){
                low[src]=min(disc[i], low[src]);
            }
        }
        return 0;
    }
    
    
    
    
    int minDays(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        
        vector<vector<int>> adj(m*n+1);
        vector<int> vis(n*m+1, 0);
        
        if(n*m==1 && grid[0][0]==1)return 1;
        if(n*m==1 && grid[0][0]==0)return 0;
        
        int x=0;
        int t = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                x++;
                if(grid[i][j]==0){continue;}
                t++;
                grid[i][j]=x; 
            }
        }
        
        if(t==1) return 1;
        
         for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(grid[i][j]!=0){
                    if(i+1<n && grid[i+1][j]!=0 ){
                        adj[grid[i][j]].push_back(grid[i+1][j]);}
                    if(j+1<m && grid[i][j+1]!=0){
                        adj[grid[i][j]].push_back(grid[i][j+1]);}
                    if(i>0 && grid[i-1][j]!=0){
                        adj[grid[i][j]].push_back(grid[i-1][j]);}
                    if(j>0 && grid[i][j-1]!=0){
                        adj[grid[i][j]].push_back(grid[i][j-1]);}

                    if(adj[grid[i][j]].size()==0){return 0;}
                }
            }
        }
        
        for(int i=0; i<=m*n; i++){
            disc[i]=0; low[i]=0; parent[i]=0;
        }
        
        int c=0;
        for(int i=1; i<=m*n; i++){
            if(vis[i]==0 && adj[i].size()!=0){
                vis[i]=1;
                c++;
                apdfs(adj,vis, i);
            }
        }
        
        if(c==1)return flag;
        return 0;
        
        
        
    }
};

File: Minimum_Number_of_Days_to_Disconnect_Island.cs
public class Solution {
    List<int>[] graph;
    int[] parent;
    bool[] visited;
    int[] desc;
    int[] low;
    bool apExists;
    int time = 0;
    
    public void APDFS(int u)
    {
        visited[u] = true;
        time++;
        desc[u] = time;
        low[u] = time;
        int children = 0;
        foreach(int v in graph[u])
        {
            if(!visited[v])
            {
                children++;
                parent[v] = u;
                APDFS(v);
                low[u] = Math.Min(low[u],low[v]);
                if(children>1 && parent[u]==-1) 
                {
                    apExists = true;
                }
                
                if(parent[u]!=-1 && low[v]>=desc[u])
                {
                    apExists = true;
                }
            }
            else
            {
                if(v!=parent[u])
                {
                    low[u] = Math.Min(low[u],desc[v]);
                }
            }
        }
    }
    
    public int MinDays(int[][] grid) {
        int n = grid.Count();
        int m = grid[0].Count();
        
        int[] xDir = new int[4]{0,0,1,-1};
        int[] yDir = new int[4]{1,-1,0,0};
        
        int V = n*m;
        graph = new List<int>[V];
        parent = new int[V];
        visited = new bool[V];
        desc = new int[V];
        low = new int[V];
        
        int landCount = 0;
               
        for(int i=0;i<V;i++)
        {
            parent[i] = -1;
            graph[i] = new List<int>();
        }
        
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==0) continue;
                landCount++;
                int u = i*m+j;
                for(int d = 0;d<4;d++)
                {
                    int x = i+xDir[d];
                    int y = j+yDir[d];
                    if(x<0 || x>=n || y<0 || y>=m || grid[x][y]==0) continue;
                    
                    int v = x*m + y;
                    graph[u].Add(v);
                }
            }
        }
        
        if(landCount<=1) return landCount;
        
        int components = 0;
        for(int u=0;u<V;u++)
        {
            int j = u%m;
            int i = (u-j)/m;
            if(!visited[u] && grid[i][j]==1)
            {
                components++;
                if(components==2) return 0;                
                APDFS(u);                
            }
        }
        
        return apExists ? 1 : 2;
    }
}